File: 2022/day02/main.swift
import CoreData
import Foundation

let input = try String(contentsOfFile: "./input.txt")
var rpsGames = input.components(separatedBy: "\n")
rpsGames.removeLast()

let decrypted = [
  "X": "A",
  "Y": "B",
  "Z": "C",
]

let scores = [
  "A": 1,
  "B": 2,
  "C": 3,
]

let rankings = [
  "A": "C",
  "B": "A",
  "C": "B",
]

let rankingsPartTwo = [
  "A": ["B", "C"],
  "B": ["C", "A"],
  "C": ["A", "B"],
]

var partOneScore = 0
var partTwoScore = 0


for game in rpsGames {
  let choices = game.components(separatedBy: " ")
  let opponentChoice = choices[0]
  let myChoice = decrypted[choices[1]]!

  // Part 1
  if opponentChoice == myChoice {
    partOneScore += 3
    partOneScore += scores[myChoice]!
  } else if rankings[opponentChoice]! == myChoice {
    partOneScore += scores[myChoice]!
  } else {
    partOneScore += 6
    partOneScore += scores[myChoice]!
  }

  // Part 2
  let myChoicePartTwo = choices[1]

  if (myChoicePartTwo == "Y") {
    partTwoScore += 3
    partTwoScore += scores[opponentChoice]!
  } else if (myChoicePartTwo == "X") { // Lost
    partTwoScore += scores[rankingsPartTwo[opponentChoice]![1]]!
  } else {
    partTwoScore += 6
    partTwoScore += scores[rankingsPartTwo[opponentChoice]![0]]!
  }
}


print("Part 1: \(partOneScore)")
print("Part 1: \(partTwoScore)")

File: 2022/day02/main.lua
local lines = {}
for line in io.lines('input.txt') do
    lines[#lines + 1] = line
end

function split(s, delimiter)
    result = {};
    for match in (s..delimiter):gmatch("(.-)"..delimiter) do
        table.insert(result, match);
    end
    return result;
end

local decrypted = {
    X="A",
    Y="B",
    Z="C",
}

local scores = {
    A=1,
    B=2,
    C=3,
}

local rankings = {
    A="C",
    B="A",
    C="B",
}

local rankingsPartTwo = {
    A={"B", "C"},
    B={"C", "A"},
    C={"A", "B"},
}

local partOneScore = 0;
local partTwoScore = 0;

for _, v in pairs(lines) do
    local choices = split(v, " ");
    local opponentsChoice = choices[1];
    local myChoice = decrypted[choices[2]];

    -- Part 1
    if (opponentsChoice == myChoice) then
        partOneScore = partOneScore + 3;
        partOneScore = partOneScore + scores[myChoice];
    elseif (rankings[opponentsChoice] == myChoice) then
        partOneScore = partOneScore + scores[myChoice];
    else
        partOneScore = partOneScore + 6;
        partOneScore = partOneScore + scores[myChoice];
    end

    -- Part 2
    local myChoicePartTwo = choices[2];

    if (myChoicePartTwo == "Y") then
        partTwoScore = partTwoScore + 3;
        partTwoScore = partTwoScore + scores[opponentsChoice];
    elseif (myChoicePartTwo == "X") then
        partTwoScore = partTwoScore + scores[rankingsPartTwo[opponentsChoice][2]];
    else
        partTwoScore = partTwoScore + 6;
        partTwoScore = partTwoScore + scores[rankingsPartTwo[opponentsChoice][1]];
    end
end

print("Part 1: "..partOneScore.."");
print ("Part 2: "..partTwoScore.."");